tests/MongoDB.Driver.Core.Tests/Core/Clusters/DnsMonitorTests.cs

```diff
@@ -332,16 +332,21 @@ public void IsValidHost_should_return_expected_result(string lookupDomainName, s
         }
 
         [Theory]
-        [InlineData(true, false)]
-        [InlineData(false, true)]
-        public void Monitor_should_return_when_ShouldDnsMonitorStop_returns_true(bool firstValue, bool secondValue)
+        [InlineData(true, false, true)]
+        [InlineData(false, true, true)]
+        [InlineData(true, null, false)]
+        public void Monitor_should_return_when_ShouldDnsMonitorStop_returns_true(bool firstValue, bool? secondValue, bool processDnsResultHasEverBeenCalled)
         {
             var mockCluster = new Mock<IDnsMonitoringCluster>();
-            mockCluster
+            var setup = mockCluster
                 .SetupSequence(m => m.ShouldDnsMonitorStop())
-                .Returns(firstValue)
-                .Returns(secondValue);
+                .Returns(firstValue);
+            if (secondValue.HasValue)
+            {
+                setup.Returns(secondValue.Value);
+            }
             var subject = CreateSubject(cluster: mockCluster.Object);
+            subject._processDnsResultHasEverBeenCalled(processDnsResultHasEverBeenCalled);
```

Comment:
I don't love how this test is reliant on changing the internal state of the test subject.

Could we instead assert that in the case where cluster.ShouldDnsMonitorStop() returns true twice in a row, that `DnsResolver.ResolveSrvRecords` (using a mock resolver) is invoked exactly once before `DnsMonitor.Subject` returns normally?

Reply:
Sorry, Jeff, I looked at the test once again and there is a problem which is related to the fact that `ShouldDnsMonitorStop` is not called first time in this condition:
    
    if (_processDnsResultHasEverBeenCalled && _cluster.ShouldDnsMonitorStop())

if `_processDnsResultHasEverBeenCalled`  is false.
So, without my change, we can only assert the second call of ShouldDnsMonitorStop

in this case, the test will validate not full expected logic in this method. Not sure that it's a good idea to change the current implementation

Reply:
I'm fine with either implementation: if Jeff's alternative isn't too much extra effort, go for it!

Reply:
Yes, because of the short-circuit &&.

I think you could write a test which expects the following

1.  A single call to DnsResolver
2. A single call to ShouldDnsMonitorStop() (which returns true)
3. Monitor method exist